README.md
# webdev

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-01-07
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-01-07
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/webdev/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/webdev/compare/0.1.0...0.1.1


src/webdev/item/view.clj
(ns webdev.item.view
  (:require [hiccup.page :refer [html5]]
            [hiccup.core :refer [html h]]))

(defn new-item []
  (html
    [:form.form-horizontal
     {:method "POST" :action "/items"}
     [:div.form-group
      [:label.control-label.col-sm-2 {:for :name-input}
       "Name"]
      [:div.col-sm-10
       [:input#name-input.form-control
        {:name :name
         :placeholder "Name"}]]]
     [:div.form-group
      [:label.control-label.col-sm-2 {:for :desc-input}
       "Description"]
      [:div.col-sm-10
       [:input#desc-input.form-control
        {:name :description
         :placeholder "Description"}]]]
     [:div.form-group
      [:div.col-sm-offset-2.col-sm-10
       [:input.btn.btn-primary
        {:type :submit
         :value "New Item"}]]]]))

(defn delete-item-form [id]
  (html
    [:form
     {:method "POST" :action (str "/items/" id)}
     [:input {:type :hidden
              :name "_method"
              :value "DELETE"}]
     [:div.btn-group
      [:input.btn.btn-danger
       {:type :submit
        :value "Delete"}]]]))

(defn update-item-form [id checked]
  (html
    [:form
     {:method "POST" :action (str "/items/" id)}
     [:input {:type :hidden
              :name "_method"
              :value "PUT"}]
     [:input {:type :hidden
              :name "checked"
              :value (if checked "false" "true")}]
     [:div.btn-group
      [:button.btn.btn-primary.btn-xs
       (if checked "Done" "Todo")]]]))

(defn items-page [items]
  (html5 {:lang :en}
         [:head
          [:title "Listronica"]
          [:meta {:name "viewport"
                  :content "width=device-width, initial-scale=1, shrink-to-fit=no"}]
          [:link {:href "/css/bootstrap.min.css"
                  :rel "stylesheet"}]
          [:body
           [:div.container
            [:h1 "My Items"]
            [:div.row
             (if (seq items)
               [:table.table.table-striped
                [:thead
                 [:tr
                  [:th]
                  [:th]
                  [:th "Name"]
                  [:th "Description"]]
                 [:tbody
                  (for [i items]
                    [:tr
                     [:td (delete-item-form (:id i))]
                     [:td (update-item-form (:id i) (:checked i))]
                     [:td (h (:name i))]
                     [:td (h (:description i))]])]]]
               [:div.col-sm-offset-1 "There are no items."])]
            [:div.col-sm-6
             [:h2 "Create a new item"]
             (new-item)]]
           [:script {:src "https://code.jquery.com/jquery-3.2.1.min.js"}]
           [:script {:src "/js/bootstrap.min.js"}]]]))

src/webdev/item/model.clj
(ns webdev.item.model
  (:require [clojure.java.jdbc :as db]))

(defn create-table
  [db]
  (db/execute!
    db
    ["CREATE EXTENSION IF NOT EXISTS \"uuid-ossp\""])
  (db/execute!
    db
    ["CREATE TABLE IF NOT EXISTS items
        (id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
         name TEXT NOT NULL,
         description TEXT NOT NULL,
         checked BOOLEAN NOT NULL DEFAULT FALSE,
         date_created TIMESTAMPTZ NOT NULL DEFAULT now())"]))

(defn create-item [db name description]
  (:id (first (db/query
                db
                ["INSERT INTO items (name, description) VALUES (?, ?) RETURNING id"
                 name
                 description]))))

(defn update-item [db id checked]
  (= [1] (db/execute!
           db
           ["UPDATE items SET checked = ? WHERE id = ?"
            checked
            id])))

(defn delete-item [db id]
  (= [1] (db/execute!
           db
           ["DELETE FROM items WHERE id = ?"
            id])))

(defn read-items [db]
   (db/query
     db
     ["SELECT id, name, description, checked, date_created FROM items ORDER BY date_created"]))

src/webdev/item/handler.clj
(ns webdev.item.handler
  (:require [webdev.item.model :refer [create-item
                                       read-items
                                       update-item
                                       delete-item]])
  (:require [webdev.item.view :refer [items-page]]))

(defn handle-index-items [req]
  (let [db (:webdev/db req)
        items (read-items db)]
    {:status 200
     :headers {}
     :body (items-page items)}))

(defn handle-create-item [req]
  (let [name (get-in req [:params "name"])
        description (get-in req [:params "description"])
        db (:webdev/db req)
        item-id (create-item db name description)]
    {:status 302
     :headers {"Location" "/items"}
     :body ""}))

(defn handle-delete-item [req]
  (let [db (:webdev/db req)
        item-id (java.util.UUID/fromString (:item-id (:route-params req)))
        exists? (delete-item db item-id)]
    (if exists?
      {:status 302
       :headers {"Location" "/items"}
       :body ""}
      {:status 404
       :body "Item not found."
       :headers {}})))

(defn handle-update-item [req]
  (let [db (:webdev/db req)
        item-id (java.util.UUID/fromString (:item-id (:route-params req)))
        checked (get-in req [:params "checked"])
        exists? (update-item db item-id (= "true" checked))]
    (if exists?
      {:status 302
       :headers {"Location" "/items"}
       :body ""}
      {:status 404
       :body "Item not found."
       :headers {}})))

src/webdev/core.clj
(ns webdev.core
  (:require [webdev.item.model :as items]
            [webdev.item.handler :refer [handle-index-items
                                         handle-create-item
                                         handle-delete-item
                                         handle-update-item]])
  (:require [ring.adapter.jetty :as jetty]
            [ring.middleware.reload :refer [wrap-reload]]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.resource :refer [wrap-resource]]
            [ring.middleware.file-info :refer [wrap-file-info]]
            [compojure.core :refer [defroutes ANY GET POST PUT DELETE]]
            [compojure.route :refer [not-found]]
            [ring.handler.dump :refer [handle-dump]]))

(def db (or
          (System/getenv "DATABASE_URL")
          {:dbtype "postgresql"
           :dbname "webdev"
           :user "postgres"}))

(defn greet [req]
  {:status 200
   :body "Hello World!"
   :headers {}})

(defn goodbye [req]
  {:status 200
   :body "Goodbye, cruel world!"
   :headers {}})

(defn about [req]
  {:status 200
   :body "My name is Caio Guedes! I created this site! Isn't it lovely?"
   :headers {}})

(defn yo
  [req]
  (let [name (get-in req [:route-params :name])]
    {:status 200
     :body (str "Yo! " name "!")
     :headers {}}))

(def ops {"+" +
          "-" -
          "*" *
          ":" /})

(defn calc
  [req]
  (let [op (get-in req [:route-params :op])
        f (get ops op)
        a (Integer. (get-in req [:route-params :a]))
        b (Integer. (get-in req [:route-params :b]))]
    (if f
      {:status 200
        :body (str (f a b))
        :headers {}}
      {:status 404
        :body "Unknown operator"
        :headers {}})))

(defroutes routes
           (GET "/" [] greet)
           (GET "/goodbye" [] goodbye)
           (GET "/about" [] about)
           (ANY "/request" [] handle-dump)
           (GET "/yo/:name" [] yo)
           (GET "/calc/:a/:op/:b" [] calc)
           (GET "/items" [] handle-index-items)
           (POST "/items" [] handle-create-item)
           (DELETE "/items/:item-id" [] handle-delete-item)
           (PUT "/items/:item-id" [] handle-update-item)
           (not-found "Page not found."))
(defn wrap-db
  [hdlr]
  (fn [req]
    (hdlr (assoc req :webdev/db db))))

(defn wrap-server
  [hdlr]
  (fn [req]
    (assoc-in (hdlr req) [:headers "Server"] "Listronica 9000")))

(def sim-methods {"PUT" :put
                 "DELETE" :delete})

(defn wrap-simulated-method [hdlr]
  (fn [req]
    (if-let [method (and (= :post (:require-method req))
                         (sim-methods (get-in req [:params "_method"])))]
      (hdlr (assoc req :request-method method))
      (hdlr req))))

(def app
  (-> routes
      (wrap-simulated-method)
      (wrap-params)
      (wrap-db)
      (wrap-resource "static")
      (wrap-file-info)
      (wrap-server)))

(defn -main [port]
  (items/create-table db)
  (jetty/run-jetty app {:port (Integer. port)}))

(defn -dev-main [port]
  (items/create-table db)
  (jetty/run-jetty (wrap-reload #'app) {:port (Integer. port)}))
